src/iqm/qiskit_iqm/iqm_provider.py

```diff
@@ -241,119 +241,148 @@ def serialize_circuit(self, circuit: QuantumCircuit) -> Circuit:
-            instruction = circuit_instruction.operation
-            qubit_names = [str(circuit.find_bit(qubit).index) for qubit in circuit_instruction.qubits]
-            if instruction.name == 'r':
-                angle_t = float(instruction.params[0] / (2 * np.pi))
-                phase_t = float(instruction.params[1] / (2 * np.pi))
-                native_inst = Instruction(name='prx', qubits=qubit_names, args={'angle_t': angle_t, 'phase_t': phase_t})
-            elif instruction.name == 'x':
-                native_inst = Instruction(name='prx', qubits=qubit_names, args={'angle_t': 0.5, 'phase_t': 0.0})
-            elif instruction.name == 'rx':
-                angle_t = float(instruction.params[0] / (2 * np.pi))
-                native_inst = Instruction(name='prx', qubits=qubit_names, args={'angle_t': angle_t, 'phase_t': 0.0})
-            elif instruction.name == 'y':
-                native_inst = Instruction(name='prx', qubits=qubit_names, args={'angle_t': 0.5, 'phase_t': 0.25})
-            elif instruction.name == 'ry':
-                angle_t = float(instruction.params[0] / (2 * np.pi))
-                native_inst = Instruction(name='prx', qubits=qubit_names, args={'angle_t': angle_t, 'phase_t': 0.25})
-            elif instruction.name == 'cz':
-                native_inst = Instruction(name='cz', qubits=qubit_names, args={})
-            elif instruction.name == 'move':
-                native_inst = Instruction(name='move', qubits=qubit_names, args={})
-            elif instruction.name == 'barrier':
-                native_inst = Instruction(name='barrier', qubits=qubit_names, args={})
-            elif instruction.name == 'measure':
-                if len(circuit_instruction.clbits) != 1:
-                    raise ValueError(
-                        f'Unexpected: measurement instruction {circuit_instruction} uses multiple classical bits.'
-                    )
-                clbit = circuit_instruction.clbits[0]  # always a single-qubit measurement
-                mk = str(MeasurementKey.from_clbit(clbit, circuit))
-                native_inst = Instruction(name='measure', qubits=qubit_names, args={'key': mk})
-                clbit_to_measure[clbit] = native_inst
-            elif instruction.name == 'reset':
-                # implemented using a measure instruction to measure the qubits, and
-                # one cc_prx per qubit to conditionally flip it to |0>
-                feedback_key = '_reset'
+        instructions = serialize_instructions(circuit, self._idx_to_qb, ignore_nonnative_gates=False)
+
+        try:
+            metadata = to_json_dict(circuit.metadata)
+        except ValueError:
+            warnings.warn(
+                f'Metadata of circuit {circuit.name} was dropped because it could not be serialised to JSON.',
+            )
+            metadata = None
+
+        return Circuit(name=circuit.name, instructions=instructions, metadata=metadata)
```

Comment:
Can we make it `serialize_circuit`, so the pure func would return a `Circuit` object (with the correct name), and after the try block it would just add metadata into the existing object?

Reply:
ok

Reply:
So it will try to revalidate the list of Instructions which already have been validated, when you construct the Circuit? How does it work now?

Reply:
Yes it validates all input Instructions recursively. I suppose it has no way of knowing they have already been validated before.

Reply:
In my current version, we never create a Circuit with invalid Instructions.